File: web/src/features/sources/reorder.ts
/**
 * The ordering primitive behind both ways of reordering this list.
 *
 * A pointer drag and an arrow-key move are the same operation with different input, so they
 * share one implementation and one set of tests. Nothing here touches the DOM, which is why
 * the keyboard path can be trusted without a browser to prove it.
 */

/** Moves one item, shifting everything between. Out-of-range or no-op moves change nothing. */
export function move<T>(items: readonly T[], from: number, to: number): T[] {
  const next = [...items]
  if (from < 0 || from >= next.length) return next
  const clamped = Math.min(Math.max(to, 0), next.length - 1)
  if (clamped === from) return next
  const [item] = next.splice(from, 1)
  next.splice(clamped, 0, item as T)
  return next
}

/** True when two orderings of the same set differ — the test for "is a save worth making". */
export function orderChanged<T>(a: readonly T[], b: readonly T[], key: (item: T) => string) {
  if (a.length !== b.length) return true
  return a.some((item, index) => key(item) !== key(b[index] as T))
}

/**
 * Where an arrow key sends the grabbed item.
 *
 * Deliberately clamped rather than wrapping: an athlete holding ArrowDown to reach the bottom
 * of the list would otherwise find their most trusted source at the top again.
 */
export function stepTo(index: number, direction: -1 | 1, length: number): number {
  return Math.min(Math.max(index + direction, 0), Math.max(length - 1, 0))
}

File: android/feature/sources/src/main/kotlin/dev/healthhub/feature/sources/Reorder.kt
package dev.healthhub.feature.sources

/**
 * The ordering primitive behind both ways of reordering the list.
 *
 * A long-press drag and a tap on the move buttons are the same operation with different input,
 * so they share one implementation and one set of tests. Nothing here touches Compose, which
 * is why the button path can be trusted without an instrumented run to prove it.
 *
 * Mirrors `web/src/features/sources/reorder.ts` rule for rule — the two clients write to the
 * same `PUT /api/sources`, so an off-by-one on one of them would show up as the phone and the
 * browser disagreeing about who is most trusted.
 */

/** Moves one item, shifting everything between. An out-of-range or no-op move changes nothing. */
internal fun <T> List<T>.movedItem(from: Int, to: Int): List<T> {
    if (from !in indices) return this
    val clamped = to.coerceIn(0, lastIndex)
    if (clamped == from) return this
    val next = toMutableList()
    next.add(clamped, next.removeAt(from))
    return next
}

/** True when two orderings of the same set differ — the test for "is a save worth making". */
internal fun <T> orderChanged(a: List<T>, b: List<T>, key: (T) -> String): Boolean {
    if (a.size != b.size) return true
    return a.indices.any { key(a[it]) != key(b[it]) }
}

/**
 * Where a move button sends the item.
 *
 * Deliberately clamped rather than wrapping: an athlete tapping "down" to reach the bottom of
 * the list would otherwise find their most trusted source back at the top.
 */
internal fun stepTo(index: Int, direction: Int, length: Int): Int =
    (index + direction).coerceIn(0, (length - 1).coerceAtLeast(0))
